src/MKL.jl

```diff
@@ -1,9 +1,53 @@
 module MKL
 
-using MKL_jll
+using Preferences
+using Libdl
 
-JULIA_VER_NEEDED = v"1.7.0-DEV.641"
-VERSION > JULIA_VER_NEEDED && using LinearAlgebra
+# Choose an MKL provider; taking an explicit preference as the first choice,
+# but if nothing is set as a preference, fall back to an environment variable,
+# and if that is not given, fall back to the default choice of `MKL_jll`.
+const mkl_provider = lowercase(something(
+    @load_preference("mkl_provider", nothing),
+    get(ENV, "JULIA_MKL_PROVIDER", nothing),
+    "mkl_jll",
+)::String)
+
+if mkl_provider == "mkl_jll"
+    # Only load MKL_jll if we are suppoed to use it as the MKL source
+    # to avoid an unnecessary download of the (lazy) artifact.
+    import MKL_jll
+    const libmkl_rt = MKL_jll.libmkl_rt
+    const mkl_path = dirname(libmkl_rt)
+elseif mkl_provider == "system"
+    # We want to use a "system" MKL, so let's try to find it.
+    # The user may provide the path to libmkl_rt via a preference
+    # or an environment variable. Otherwise, we expect it to
+    # already be loaded, or be on our linker search path.
+    const mkl_path = lowercase(something(
+        @load_preference("mkl_path", nothing),
+        get(ENV, "JULIA_MKL_PATH", nothing),
+        "",
+    )::String)
+    libname = string("libmkl_rt", ".", Libdl.dlext)
+    const libmkl_rt = find_library(libname, [mkl_path])
+    libmkl_rt == "" && error("Couldn't find $libname. Maybe try setting JULIA_MKL_PATH?")
+else
+    error("Invalid mkl_provider choice $(mkl_provider).")
+end
+
+# Changing the MKL provider preference
+function set_mkl_provider(provider)
+    if lowercase(provider) ∉ ("mkl_jll", "system")
+        error("Invalid mkl_provider choice $(provider)")
+    end
```

Comment:
If we combine the two, this check should basically check to make sure that the file exists.